PP-915: pbs_snapshot doesn't anonymize Authorized_Users & Authorized_…

agrawalravi90 opened the pull request:
…Groups

<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-915](https://pbspro.atlassian.net/browse/PP-915)**

#### Problem description
* pbs_snapshot doesn't anonymize Authorized_Users & Authorized_Groups

#### Cause / Analysis
* I'd forgotten to add Authorized_Users & Authorized_Groups to the list of attributes to obfuscate.

#### Solution description
* Added Authorized_Users & Authorized_Groups to the list of attributes to obfuscate.
    * Modified pbs_anonutils to deal with these two attributes as lists.
* Also added Reserve_Owner to the list of attributes to obfuscate.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

arungrover commented on test/tests/functional/pbs_snapshot_unittest.py:
```diff
@@ -300,3 +302,62 @@ def test_snapshot_without_logs(self):
+        self.server.expect(RESV, attribs, id=resv_id)
+
+        # Now, take a snapshot with --obfuscate
+        target_dir = os.getcwd()
```
can this be tmp which in most cases is likely to pass for any user that is going to run this PTL script.

agrawalravi90 replied:
it could be, yes, I'll change it to that, thanks.

arungrover commented on test/tests/functional/pbs_snapshot_unittest.py:
```diff
@@ -300,3 +302,62 @@ def test_snapshot_without_logs(self):
+        # Authorized user and group names
+        pbsrstat_path = os.path.join(snap_dir, PBS_RSTAT_F_PATH)
+        self.assertTrue(os.path.isfile(pbsrstat_path))
+        with open(pbsrstat_path, "r") as rstatfd:
```
Overall change looks good. I have a nit about what happens to the test case if pbsrstat_path fails to open. I think you may way to fail the test in that case.

I also have a general question... when you execute snapshot as a super user, what are the permissions of the tarball created in target directory? If I run this PTL script as a normal user will it be able to read the snapshot? If not then we should make sure that the script is always run as root user. what do you think?

agrawalravi90 replied:
For what happens if pbsrstat_path fails to open, the with statement will raise an IOError, which, I believe, will fail the test case.

"If I run this PTL script as a normal user will it be able to read the snapshot?"
hmm .. you bring up a good point. pbs_snapshot only supports being run with root privileges. So, we should probably skip all the tests in this testsuite if this is run by a user who doesn't have root privileges. I'll change the testsuite to do that.

agrawalravi90 commented:
Hey Arun,

I've addressed your comments, please take a look the changes.

About the root access thing, I've added the code to skip the test suite if the user doesn't have root access. While testing to see if the skip actually works, when I ran these tests as a non-privileged user, I came across this:

**_2017-08-15 19:45:18,474 INFO     expect action: added action kicksched to server pbspro
[sudo] password for normal_user: 
[sudo] password for normal_user: 
[sudo] password for normal_user: 
2017-08-15 19:45:39,200 ERROR    err: ['', 'We trust you have received the usual lecture from the local System', 'Administrator. It usually boils down to these three things:', '', '    #1) Respect the privacy of others.', '    #2) Think before you type.', '    #3) With great power comes great responsibility.', '', 'Sorry, try again.', 'Sorry, try again.', 'Sorry, try again.', 'sudo: 3 incorrect password attempts']
2017-08-15 19:45:39,202 INFOCLI2 pbspro: sudo -H /opt/pbs/sbin/pbsfs_**

I might be wrong, but it seems like one cannot run PTL tests at all without having root privileges? It error'd out much before it could reach my code to skip the tests. Please let me know what you think.

arungrover commented on test/tests/functional/pbs_snapshot_unittest.py:
```diff
@@ -55,6 +58,24 @@ def setUp(self):
+            if ret['rc'] == 0:
+                self.pbs_snapshot_exists = True
+            else:
+                self.pbs_snapshot_exists = False
```
where do you use this flag?

agrawalravi90 replied:
Ah, thanks for catching that, I'd meant to add that to skip the tests which use pbs_snapshot binary, if the binary is not installed. I've changed the code to reflect that now. Also added logs to show how this will work to the ticket. Please review the changes once more. Thanks!

arungrover commented:
Looks good @agrawalravi90. Once you have another sign-off please rebase and squash

agrawalravi90 commented:
Thanks Arun.

@PBSPro/pbspro-maintainers requesting a maintainer to take this up. Thanks!

agrawalravi90 commented:
Thanks Mike, I just squashed and rebased.
